### PR TITLE
chore: Optimize getConfig

### PR DESCRIPTION
**Pull request by jimblanc**
#### Description of changes

This change optimizes `getConfig` by generating immutable resource config at configure-time and updates it to return an immutable object.

#### Description of how you validated changes
Unit tests, smoke testing in sample app.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by stocaaro — APPROVED** [review]

**Review by israx — APPROVED** [review]
LGTM!

File: packages/core/src/utils/objectHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (cshfang) [inline packages/core/src/utils/objectHelpers.ts:1]: Nit: Maybe just name this file `deepFreeze.ts` unless we're certain there will be more object helpers?

**Review by cshfang — APPROVED** [review]

File: packages/core/src/singleton/Amplify.ts [hunk 2]
```diff
@@ -71,10 +73,10 @@ export class AmplifyClass {
 	/**
 	 * Provides access to the current back-end resource configuration for the Library.
 	 *
-	 * @returns Returns the current back-end resource configuration.
+	 * @returns Returns the immutable back-end resource configuration.
 	 */
 	getConfig(): ResourcesConfig {
```
Comment (cshfang) [inline packages/core/src/singleton/Amplify.ts:78]: Should this be `ReadOnly<ResourcesConfig>`?

**Review by cshfang — APPROVED** [review]

**Review by cshfang — APPROVED** [review]

File: packages/core/src/utils/deepFreeze.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+export const deepFreeze = (object: any) => {
+	const propNames = Reflect.ownKeys(object);
```
Comment (elorzafe) [inline packages/core/src/utils/deepFreeze.ts:5]: This is well supported for Web https://caniuse.com/?search=Reflect

Can we say the same for node and react-native? Did you tests this on Next.js?

Comment (jimblanc) [reply packages/core/src/utils/deepFreeze.ts:5]: Supported at least back to Node 14 & tested in a Next.js sample app. Reflect is standard ES6 & supported in RN, confirmed with sample app.

**Pull request merged**